compositor_pipeline/src/pipeline/whip_whep.rs

```diff
@@ -0,0 +1,194 @@
+use crate::pipeline::input::whip::depayloader::Depayloader;
+use axum::{
+    routing::{delete, get, patch, post},
+    Router,
+};
+use compositor_render::InputId;
+use error::WhipServerError;
+use reqwest::StatusCode;
+use serde_json::{json, Value};
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    sync::{Arc, Mutex},
+    time::{Duration, Instant},
+};
+use tokio::{
+    signal,
+    sync::{mpsc, oneshot},
+};
+use tower_http::cors::CorsLayer;
+use tracing::{error, warn};
+use webrtc::{
+    peer_connection::{peer_connection_state::RTCPeerConnectionState, RTCPeerConnection},
+    rtp_transceiver::rtp_codec::RTPCodecType,
+};
+use whip_handlers::{
+    create_whip_session::handle_create_whip_session,
+    new_whip_ice_candidates::handle_new_whip_ice_candidates,
+    terminate_whip_session::handle_terminate_whip_session,
+};
+
+mod error;
+mod init_peer_connection;
+mod validate_bearer_token;
+mod whip_handlers;
+
+use crate::queue::PipelineEvent;
+
+use super::EncodedChunk;
+
+pub async fn run_whip_whep_server(
+    port: u16,
+    state: Arc<WhipWhepState>,
+    shutdown_signal_receiver: oneshot::Receiver<()>,
+) {
+    let app = Router::new()
+        .route("/status", get(status))
+        .route("/whip/:id", post(handle_create_whip_session))
+        .route("/session/:id", patch(handle_new_whip_ice_candidates))
+        .route("/session/:id", delete(handle_terminate_whip_session))
+        .layer(CorsLayer::permissive())
+        .with_state(state.clone());
+
+    let Ok(listener) = tokio::net::TcpListener::bind(SocketAddr::from(([0, 0, 0, 0], port))).await
+    else {
+        warn!("TCP listener error");
+        return;
+    };
+
+    if let Err(err) = axum::serve(listener, app)
+        .with_graceful_shutdown(shutdown_signal(shutdown_signal_receiver))
+        .await
+    {
+        error!("Cannot serve WHIP/WHEP server task: {err:?}");
+    };
+    state.clear_state();
+}
+
+async fn shutdown_signal(receiver: oneshot::Receiver<()>) {
+    let ctrl_c = async {
+        signal::ctrl_c()
+            .await
+            .expect("failed to install Ctrl+C handler");
+    };
+
+    let terminate = async {
+        if let Err(err) = receiver.await {
+            warn!(
+                "Error while receiving whip_whep server shutdown signal {:?}",
+                err
+            );
+        }
+    };
+
+    tokio::select! {
+        _ = ctrl_c => {},
+        _ = terminate => {},
+    }
+}
+
+pub async fn status() -> Result<(StatusCode, axum::Json<Value>), WhipServerError> {
+    Ok((StatusCode::OK, axum::Json(json!({}))))
```

Comment:
feels like something that could be inline, but I don't have any strong opinion on that